File: desktop-app-vue/src/main/llm/anthropic-client.js
/**
 * Anthropic Claude API client.
 * Supports chat and streaming via the Messages API.
 */

const { logger, createLogger } = require('../utils/logger.js');
const axios = require('axios');
const EventEmitter = require('events');

class AnthropicClient extends EventEmitter {
  constructor(config = {}) {
    super();

    this.apiKey = config.apiKey;
    this.baseURL = config.baseURL || 'https://api.anthropic.com';
    this.model = config.model || 'claude-3-opus-20240229';
    this.timeout = config.timeout || 120000;
    this.maxTokens = config.maxTokens || 2000;
    this.anthropicVersion = config.anthropicVersion || '2023-06-01';

    this.client = axios.create({
      baseURL: this.baseURL,
      timeout: this.timeout,
      headers: {
        'Content-Type': 'application/json',
        ...(this.apiKey && { 'x-api-key': this.apiKey }),
        'anthropic-version': this.anthropicVersion,
      },
    });
  }

  normalizeMessages(messages = [], options = {}) {
    let systemPrompt = '';
    let sawSystemMessage = false;
    const normalized = [];

    for (const message of messages) {
      if (!message || !message.role) {continue;}

      const content =
        typeof message.content === 'string'
          ? message.content
          : JSON.stringify(message.content ?? '');

      if (message.role === 'system') {
        sawSystemMessage = true;
        systemPrompt = systemPrompt ? `${systemPrompt}\n${content}` : content;
      } else if (message.role === 'user' || message.role === 'assistant') {
        normalized.push({ role: message.role, content });
      }
    }

    if (!sawSystemMessage && options.systemPrompt) {
      systemPrompt = options.systemPrompt;
    }

    return {
      system: systemPrompt || undefined,
      messages: normalized,
    };
  }

  buildPayload(messages, options = {}, stream = false) {
    const { system, messages: normalized } = this.normalizeMessages(messages, options);
    const maxTokens = options.max_tokens || options.maxTokens || this.maxTokens;

    const payload = {
      model: options.model || this.model,
      max_tokens: maxTokens,
      messages: normalized,
      stream,
    };

    if (system) {payload.system = system;}
    if (options.temperature !== undefined) {payload.temperature = options.temperature;}
    if (options.top_p !== undefined) {payload.top_p = options.top_p;}
    if (options.top_k !== undefined) {payload.top_k = options.top_k;}
    if (Array.isArray(options.stop_sequences) && options.stop_sequences.length > 0) {
      payload.stop_sequences = options.stop_sequences;
    }

    return payload;
  }

  async checkStatus() {
    if (!this.apiKey) {
      return {
        available: false,
        error: 'Anthropic API key is not configured',
        models: [],
      };
    }

    try {
      const response = await this.client.get('/v1/models');
      const models = response.data?.data || [];

      return {
        available: true,
        models: models.map((m) => ({
          name: m.id || m.name || m.model,
        })),
      };
    } catch (error) {
      return {
        available: false,
        error: error.response?.data?.error?.message || error.message,
        models: [],
      };
    }
  }

  async chat(messages, options = {}) {
    try {
      const payload = this.buildPayload(messages, options, false);
      const response = await this.client.post('/v1/messages', payload);
      const blocks = response.data?.content || [];
      const text = blocks
        .filter((block) => block.type === 'text')
        .map((block) => block.text)
        .join('');

      const usage = response.data?.usage || {};

      return {
        message: { role: 'assistant', content: text },
        text,
        model: response.data?.model || payload.model,
        usage,
        tokens: (usage.input_tokens || 0) + (usage.output_tokens || 0),
      };
    } catch (error) {
      logger.error('[AnthropicClient] chat failed:', error.response?.data || error);
      throw new Error(error.response?.data?.error?.message || error.message);
    }
  }

  async chatStream(messages, onChunk, options = {}) {
    try {
      const payload = this.buildPayload(messages, options, true);
      const response = await this.client.post('/v1/messages', payload, {
        responseType: 'stream',
      });

      const fullMessage = { role: 'assistant', content: '' };
      let buffer = '';
      let settled = false;

      const maybeResolve = (finishReason = 'stop', resolve) => {
        if (settled) {return;}
        settled = true;
        resolve({
          message: fullMessage,
          text: fullMessage.content,
          model: payload.model,
          finish_reason: finishReason,
        });
      };

      const handleEvent = (event, data, resolve) => {
        if (!data) {return;}
        if (data === '[DONE]') {
          maybeResolve('stop', resolve);
          return;
        }

        let parsed;
        try {
          parsed = JSON.parse(data);
        } catch (e) {
          return;
        }

        if (event === 'content_block_start') {
          const text = parsed.content_block?.text;
          if (text) {
            fullMessage.content += text;
            if (typeof onChunk === 'function') {
              onChunk(text, fullMessage.content);
            }
          }
        } else if (event === 'content_block_delta') {
          const text = parsed.delta?.text;
          if (text) {
            fullMessage.content += text;
            if (typeof onChunk === 'function') {
              onChunk(text, fullMessage.content);
            }
          }
        } else if (event === 'message_stop') {
          maybeResolve('stop', resolve);
        } else if (event === 'message_delta' && parsed.delta?.stop_reason) {
          maybeResolve(parsed.delta.stop_reason, resolve);
        }
      };

      return await new Promise((resolve, reject) => {
        const processBuffer = () => {
          buffer = buffer.replace(/\r\n/g, '\n');
          let boundary = buffer.indexOf('\n\n');
          while (boundary !== -1) {
            const rawEvent = buffer.slice(0, boundary).trim();
            buffer = buffer.slice(boundary + 2);
            boundary = buffer.indexOf('\n\n');

            if (!rawEvent) {continue;}

            const lines = rawEvent.split('\n');
            let event = '';
            const dataLines = [];

            for (const line of lines) {
              if (line.startsWith('event:')) {
                event = line.slice('event:'.length).trim();
              } else if (line.startsWith('data:')) {
                dataLines.push(line.slice('data:'.length).trim());
              }
            }

            if (dataLines.length === 0) {continue;}
            const data = dataLines.join('\n');
            handleEvent(event, data, resolve);
          }
        };

        response.data.on('data', (chunk) => {
          buffer += chunk.toString();
          processBuffer();
        });

        response.data.on('error', (error) => {
          if (settled) {return;}
          settled = true;
          reject(error);
        });

        response.data.on('end', () => {
          maybeResolve('stop', resolve);
        });
      });
    } catch (error) {
      logger.error('[AnthropicClient] stream chat failed:', error.response?.data || error);
      throw new Error(error.response?.data?.error?.message || error.message);
    }
  }

  async embeddings() {
    throw new Error('Anthropic does not support embeddings.');
  }
}

module.exports = {
  AnthropicClient,
};

File: ios-app/ChainlessChain/Features/AI/Services/AnthropicClient.swift
import Foundation

/// Anthropic (Claude) API Client
class AnthropicClient: LLMClient {
    private let apiKey: String?
    private let baseURL: String
    private let model: String
    private let timeout: TimeInterval
    private let anthropicVersion: String
    private let maxTokens: Int
    private let session: URLSession

    init(config: LLMManager.LLMConfig) {
        self.apiKey = config.apiKey
        self.baseURL = config.baseURL == "http://localhost:11434" ? "https://api.anthropic.com/v1" : config.baseURL
        self.model = config.model == "qwen2:7b" ? "claude-3-5-sonnet-20241022" : config.model
        self.timeout = config.timeout
        self.anthropicVersion = "2023-06-01"
        self.maxTokens = 4096

        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = timeout
        configuration.timeoutIntervalForResource = timeout
        self.session = URLSession(configuration: configuration)
    }

    func checkStatus() async throws -> ServiceStatus {
        // Anthropic doesn't have a models endpoint, return predefined models
        let models = [
            "claude-3-5-sonnet-20241022",
            "claude-3-5-haiku-20241022",
            "claude-3-opus-20240229",
            "claude-3-sonnet-20240229",
            "claude-3-haiku-20240307"
        ]

        return ServiceStatus(available: apiKey != nil, models: models, error: apiKey == nil ? "API key required" : nil)
    }

    func chat(messages: [LLMMessage], options: ChatOptions) async throws -> LLMResponse {
        guard let url = URL(string: "\(baseURL)/messages") else {
            throw LLMError.networkError("Invalid URL")
        }

        // Convert messages to Anthropic format
        let (system, anthropicMessages) = convertMessages(messages)

        let requestBody = MessagesRequest(
            model: options.model ?? model,
            max_tokens: options.maxTokens ?? maxTokens,
            temperature: options.temperature,
            top_p: options.topP,
            system: system,
            messages: anthropicMessages,
            stream: false
        )

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(anthropicVersion, forHTTPHeaderField: "anthropic-version")
        if let apiKey = apiKey {
            request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        }
        request.httpBody = try JSONEncoder().encode(requestBody)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            // Try to parse error message
            if let errorResponse = try? JSONDecoder().decode(ErrorResponse.self, from: data) {
                throw LLMError.apiError(errorResponse.error.message)
            }
            throw LLMError.apiError("HTTP error: \((response as? HTTPURLResponse)?.statusCode ?? 0)")
        }

        let messagesResponse = try JSONDecoder().decode(MessagesResponse.self, from: data)

        // Extract text from content blocks
        let text = messagesResponse.content.compactMap { block in
            if case .text(let textBlock) = block {
                return textBlock.text
            }
            return nil
        }.joined()

        return LLMResponse(
            text: text,
            model: messagesResponse.model,
            tokens: messagesResponse.usage.input_tokens + messagesResponse.usage.output_tokens,
            usage: LLMResponse.TokenUsage(
                promptTokens: messagesResponse.usage.input_tokens,
                completionTokens: messagesResponse.usage.output_tokens,
                totalTokens: messagesResponse.usage.input_tokens + messagesResponse.usage.output_tokens
            ),
            timestamp: Date()
        )
    }

    func chatStream(
        messages: [LLMMessage],
        options: ChatOptions,
        onChunk: @escaping (String) -> Void
    ) async throws -> LLMResponse {
        guard let url = URL(string: "\(baseURL)/messages") else {
            throw LLMError.networkError("Invalid URL")
        }

        // Convert messages to Anthropic format
        let (system, anthropicMessages) = convertMessages(messages)

        let requestBody = MessagesRequest(
            model: options.model ?? model,
            max_tokens: options.maxTokens ?? maxTokens,
            temperature: options.temperature,
            top_p: options.topP,
            system: system,
            messages: anthropicMessages,
            stream: true
        )

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(anthropicVersion, forHTTPHeaderField: "anthropic-version")
        if let apiKey = apiKey {
            request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        }
        request.httpBody = try JSONEncoder().encode(requestBody)

        let (asyncBytes, response) = try await session.bytes(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw LLMError.apiError("HTTP error: \((response as? HTTPURLResponse)?.statusCode ?? 0)")
        }

        var fullText = ""
        var lastModel: String?
        var inputTokens = 0
        var outputTokens = 0

        for try await line in asyncBytes.lines {
            guard line.hasPrefix("data: ") else { continue }

            let data = line.dropFirst(6) // Remove "data: " prefix

            guard let jsonData = data.data(using: .utf8),
                  let event = try? JSONDecoder().decode(StreamEvent.self, from: jsonData) else {
                continue
            }

            switch event.type {
            case "content_block_delta":
                if let delta = event.delta,
                   case .textDelta(let textDelta) = delta {
                    fullText += textDelta.text
                    onChunk(textDelta.text)
                }

            case "message_start":
                if let message = event.message {
                    lastModel = message.model
                    inputTokens = message.usage.input_tokens
                }

            case "message_delta":
                if let usage = event.usage {
                    outputTokens = usage.output_tokens
                }

            case "message_stop":
                break

            default:
                break
            }
        }

        return LLMResponse(
            text: fullText,
            model: lastModel,
            tokens: inputTokens + outputTokens,
            usage: LLMResponse.TokenUsage(
                promptTokens: inputTokens,
                completionTokens: outputTokens,
                totalTokens: inputTokens + outputTokens
            ),
            timestamp: Date()
        )
    }

    // MARK: - Helper Methods

    private func convertMessages(_ messages: [LLMMessage]) -> (String?, [AnthropicMessage]) {
        var system: String?
        var anthropicMessages: [AnthropicMessage] = []

        for message in messages {
            if message.role == "system" {
                system = message.content
            } else {
                anthropicMessages.append(AnthropicMessage(
                    role: message.role,
                    content: [.text(TextContent(text: message.content))]
                ))
            }
        }

        return (system, anthropicMessages)
    }

    // MARK: - Request/Response Types

    private struct MessagesRequest: Codable {
        let model: String
        let max_tokens: Int
        let temperature: Double?
        let top_p: Double?
        let system: String?
        let messages: [AnthropicMessage]
        let stream: Bool
    }

    private struct AnthropicMessage: Codable {
        let role: String
        let content: [ContentBlock]
    }

    private enum ContentBlock: Codable {
        case text(TextContent)

        enum CodingKeys: String, CodingKey {
            case type, text
        }

        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            let type = try container.decode(String.self, forKey: .type)

            switch type {
            case "text":
                let text = try container.decode(String.self, forKey: .text)
                self = .text(TextContent(text: text))
            default:
                throw DecodingError.dataCorruptedError(forKey: .type, in: container, debugDescription: "Unknown content type")
            }
        }

        func encode(to encoder: Encoder) throws {
            var container = encoder.container(keyedBy: CodingKeys.self)

            switch self {
            case .text(let textContent):
                try container.encode("text", forKey: .type)
                try container.encode(textContent.text, forKey: .text)
            }
        }
    }

    private struct TextContent: Codable {
        let text: String
    }

    private struct MessagesResponse: Codable {
        let id: String
        let type: String
        let role: String
        let content: [ContentBlock]
        let model: String
        let stop_reason: String?
        let usage: Usage

        struct Usage: Codable {
            let input_tokens: Int
            let output_tokens: Int
        }
    }

    private struct StreamEvent: Codable {
        let type: String
        let message: MessageInfo?
        let delta: Delta?
        let usage: UsageInfo?

        struct MessageInfo: Codable {
            let model: String
            let usage: UsageInfo
        }

        struct UsageInfo: Codable {
            let input_tokens: Int
            let output_tokens: Int
        }

        enum Delta: Codable {
            case textDelta(TextDelta)

            struct TextDelta: Codable {
                let text: String
            }

            enum CodingKeys: String, CodingKey {
                case type, text
            }

            init(from decoder: Decoder) throws {
                let container = try decoder.container(keyedBy: CodingKeys.self)
                let type = try container.decode(String.self, forKey: .type)

                switch type {
                case "text_delta":
                    let text = try container.decode(String.self, forKey: .text)
                    self = .textDelta(TextDelta(text: text))
                default:
                    throw DecodingError.dataCorruptedError(forKey: .type, in: container, debugDescription: "Unknown delta type")
                }
            }

            func encode(to encoder: Encoder) throws {
                var container = encoder.container(keyedBy: CodingKeys.self)

                switch self {
                case .textDelta(let textDelta):
                    try container.encode("text_delta", forKey: .type)
                    try container.encode(textDelta.text, forKey: .text)
                }
            }
        }
    }

    private struct ErrorResponse: Codable {
        let error: ErrorDetail

        struct ErrorDetail: Codable {
            let message: String
            let type: String
        }
    }
}
